Redact sensitive information from log output

ethernetdan opened the pull request:
Adds option to remove a list of strings from log output. Both access and refresh tokens are automatically included.

External users can specify their own strings to be redacted using the log builders or `AddRedactStrs` once a logger has been created.

/cc @jhernand

jhernand reviewed (changes requested):
The `Redactor` idea is nice. However I have some concerns.

First is that it means that we increase the public interface of the SDK, it becomes a contract that we have to support going forward. I'd prefer to implement this without making the details part of the contract.

My second concern is that storing the redacted strings in memory is a potential source of both memory leaks, security issues and concurrency problems. In a long running application tokens are renewed every few minutes, and we would be adding them to the `redactStrs` map, and never removing them. This is a potential performance problem, because it is a memory leak. It is also a potential security problem because all the history of tokens would be stored in memory, in dumps, in swap, etc. It is also problematic for concurrency, because maps aren't safe to use by multiple goroutines simultaneously. All these problems can be solved using a concurrent map, and adding cleanup code. But I think it makes it more complicated and fragile than I would like it to be.

My preference would be to do only what you did with `redactFields` and `redactSensitive`, but instead of using the redactor concept just modify the JSON document to replace the sensitive fields with `***` before dumping. This will also require a change in the `Connection.sendTokenForm` method, to also avoid printing the refresh token sent to the server.

I think that with that we will have a reasonably sane log output, eliminating completely the concerns about memory leaks, security and concurrency.

ethernetdan commented:
Those are definitely reasonable concerns, I'll simplify it modify values directly before printing as suggested.

Commit:
Redact sensitive fields.

Removes senitive fields when request and response data are dumped to
a logger. The slice redactFields contains the names of removed fields.

jhernand commented:
Looks good, thanks @ethernetdan.
